File: ios/RBDarts/Shared/Data/RBDartsRepository.swift
import Foundation

public protocol RBDartsRepository: Sendable {
    func createStandaloneGame(_ game: StandaloneGame, participants: [ParticipantScorecard]) async throws
    func saveStandaloneSummary(_ summary: GameSummary, for gameId: String) async throws
    func createLeague(_ league: League) async throws
    func createMatch(_ match: Match) async throws
    func savePracticeAttempt(_ attempt: PracticeAttempt) async throws
    func requestCorrection(_ correction: CorrectionAuditRecord) async throws
}

public actor InMemoryRBDartsRepository: RBDartsRepository {
    public private(set) var standaloneGames: [String: StandaloneGame] = [:]
    public private(set) var standaloneSummaries: [String: GameSummary] = [:]
    public private(set) var leagues: [String: League] = [:]
    public private(set) var matches: [String: Match] = [:]
    public private(set) var practiceAttempts: [PracticeAttempt] = []
    public private(set) var corrections: [CorrectionAuditRecord] = []

    public init() {}

    public func createStandaloneGame(_ game: StandaloneGame, participants: [ParticipantScorecard]) {
        standaloneGames[game.standaloneGameId] = game
    }

    public func saveStandaloneSummary(_ summary: GameSummary, for gameId: String) {
        standaloneSummaries[gameId] = summary
    }

    public func createLeague(_ league: League) {
        leagues[league.leagueId] = league
    }

    public func createMatch(_ match: Match) {
        matches[match.matchId] = match
    }

    public func savePracticeAttempt(_ attempt: PracticeAttempt) {
        practiceAttempts.append(attempt)
    }

    public func requestCorrection(_ correction: CorrectionAuditRecord) {
        corrections.append(correction)
    }
}

File: android/app/src/main/java/com/rbdarts/core/data/RBDartsRepository.kt
package com.rbdarts.core.data

import com.rbdarts.core.domain.CorrectionAuditRecord
import com.rbdarts.core.domain.GameSummary
import com.rbdarts.core.domain.League
import com.rbdarts.core.domain.Match
import com.rbdarts.core.domain.ParticipantScorecard
import com.rbdarts.core.domain.PracticeAttempt
import com.rbdarts.core.domain.StandaloneGame

interface RBDartsRepository {
    suspend fun createStandaloneGame(game: StandaloneGame, participants: List<ParticipantScorecard>)
    suspend fun saveStandaloneSummary(summary: GameSummary, gameId: String)
    suspend fun createLeague(league: League)
    suspend fun createMatch(match: Match)
    suspend fun savePracticeAttempt(attempt: PracticeAttempt)
    suspend fun requestCorrection(correction: CorrectionAuditRecord)
}

class InMemoryRBDartsRepository : RBDartsRepository {
    val standaloneGames = linkedMapOf<String, StandaloneGame>()
    val standaloneSummaries = linkedMapOf<String, GameSummary>()
    val leagues = linkedMapOf<String, League>()
    val matches = linkedMapOf<String, Match>()
    val practiceAttempts = mutableListOf<PracticeAttempt>()
    val corrections = mutableListOf<CorrectionAuditRecord>()

    override suspend fun createStandaloneGame(game: StandaloneGame, participants: List<ParticipantScorecard>) {
        standaloneGames[game.standaloneGameId] = game
    }

    override suspend fun saveStandaloneSummary(summary: GameSummary, gameId: String) {
        standaloneSummaries[gameId] = summary
    }

    override suspend fun createLeague(league: League) {
        leagues[league.leagueId] = league
    }

    override suspend fun createMatch(match: Match) {
        matches[match.matchId] = match
    }

    override suspend fun savePracticeAttempt(attempt: PracticeAttempt) {
        practiceAttempts += attempt
    }

    override suspend fun requestCorrection(correction: CorrectionAuditRecord) {
        corrections += correction
    }
}
